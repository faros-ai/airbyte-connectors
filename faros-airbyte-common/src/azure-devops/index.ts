export * from './types';
export {AzureDevOps} from './azure-devops';
export * from './utils';
