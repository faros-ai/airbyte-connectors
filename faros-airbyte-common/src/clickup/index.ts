export interface Workspace {
  id: string;
  name: string;
  color: string;
  avatar: string;
  members: readonly Member[];
}

interface Member {
  user: User;
}

interface User {
  id: number;
  username: string;
  email: string;
  color: string;
  profilePicture: string;
  initials: string;
  role: number;
  last_active: string;
  date_joined: string;
  date_invited: string;
}

export interface Space {
  computedProperties: {workspace: {id: string}};
  id: string;
  name: string;
  private: boolean;
  statuses: ReadonlyArray<Status>;
  multiple_assignees: boolean;
  features: Features;
}

interface Features {
  due_dates: DueDates;
  time_tracking: Checklists;
  tags: Checklists;
  time_estimates: Checklists;
  checklists: Checklists;
  custom_fields: Checklists;
  remap_dependencies: Checklists;
  dependency_warning: Checklists;
  portfolios: Checklists;
}

interface Checklists {
  enabled: boolean;
}

interface DueDates {
  enabled: boolean;
  start_date: boolean;
  remap_due_dates: boolean;
  remap_closed_due_date: boolean;
}

export interface Folder {
  computedProperties: {workspace: {id: string}};
  id: string;
  name: string;
  orderindex: number;
  override_statuses: boolean;
  hidden: boolean;
  space: {id: string; name: string; access: boolean};
  task_count: string;
  lists: readonly any[];
}

export interface List {
  computedProperties: {workspace: {id: string}};
  id: string;
  name: string;
  orderindex: number;
  content: string;
  status: {status: string; color: string; hide_label: boolean};
  priority: {priority: string; color: string};
  assignee: null;
  task_count: null;
  due_date: string;
  start_date: null;
  folder: {id: string; name: string; hidden?: boolean; access: boolean};
  space: {id: string; name: string; access: boolean};
  archived: boolean;
  override_statuses: boolean;
  permission_level: string;
}

interface Status {
  status: string;
  color: string;
  orderindex: number;
  type: string;
}

export interface Task {
  computedProperties: {workspace: {id: string}};
  id: string;
  custom_id: null;
  name: string;
  text_content: string;
  description: string;
  status: Status;
  orderindex: string;
  date_created: string;
  date_updated: string;
  date_closed: string;
  archived: boolean;
  creator: Creator;
  assignees: readonly Creator[];
  watchers: readonly Creator[];
  checklists: readonly Checklist[];
  tags: readonly Tag[];
  parent: string;
  priority: number;
  due_date: string;
  start_date: string;
  points: number;
  time_estimate: number;
  custom_fields: readonly CustomField[];
  dependencies: readonly any[];
  linked_tasks: readonly any[];
  team_id: string;
  url: string;
  permission_level: string;
  list: FolderRef;
  project: FolderRef;
  folder: FolderRef;
  space: SpaceRef;
}

interface Creator {
  id: number;
  username: string;
  color: string;
  email: string;
  profilePicture: string;
}

interface Checklist {
  id: string;
  task_id: string;
  name: string;
  date_created: string;
  orderindex: number;
  creator: number;
  resolved: number;
  unresolved: number;
  items: readonly Item[];
}

interface Item {
  id: string;
  name: string;
  orderindex: number;
  assignee: null;
  resolved: boolean;
  parent: null;
  date_created: string;
  children: readonly any[];
}

interface CustomField {
  id: string;
  name: string;
  type: string;
  type_config: Record<string, unknown>;
  date_created: string;
  hide_from_guests: boolean;
  required: boolean;
}

interface FolderRef {
  id: string;
  name: string;
  hidden?: boolean;
  access: boolean;
}

interface SpaceRef {
  id: string;
}

interface Tag {
  name: string;
  tag_fg: string;
  tag_bg: string;
}

export interface Goal {
  id: string;
  name: string;
  team_id: string;
  date_created: string;
  start_date: null;
  due_date: string;
  description: string;
  private: boolean;
  archived: boolean;
  creator: number;
  color: string;
  pretty_id: string;
  multiple_owners: boolean;
  folder_id: null;
  members: readonly any[];
  owners: readonly Owner[];
  key_results: readonly any[];
  percent_completed: number;
  history: readonly any[];
  pretty_url: string;
}

interface Owner {
  id: number;
  username: string;
  initials: string;
  email: string;
  color: string;
  profilePicture: string;
}
