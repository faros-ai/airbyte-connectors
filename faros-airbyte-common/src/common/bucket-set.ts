import VError from 'verror';

export class BucketSet {
  private readonly sortedBuckets: ReadonlyArray<number>;

  constructor(
    bucketTotal: number,
    bucketRanges: string | ReadonlyArray<string>
  ) {
    if (bucketTotal < 1) {
      throw new VError('bucket_total must be a positive integer');
    }

    const rangeStrings = (
      typeof bucketRanges === 'string' ? bucketRanges.split(',') : bucketRanges
    ).filter((s) => s.trim());

    if (!rangeStrings?.length) {
      throw new VError('bucket_ranges cannot be empty');
    }

    const buckets = new Set<number>();

    for (const range of rangeStrings) {
      const [start, end] = this.parseRange(range);

      if (start < 1 || end > bucketTotal) {
        throw new VError(
          `Invalid bucket range ${range}: values must be between 1 and ${bucketTotal}`
        );
      }

      for (let i = start; i <= end; i++) {
        buckets.add(i);
      }
    }

    this.sortedBuckets = Array.from(buckets).sort((a, b) => a - b);
  }

  private parseRange(range: string): [number, number] {
    const parts = range.split('-').map((p) => p.trim());

    if (parts.length > 2) {
      throw new VError(
        `Invalid range format: ${range}. Valid formats are: single number (e.g., '7') or number range (e.g., '3-5')`
      );
    }

    const start = parseInt(parts[0]);
    if (isNaN(start)) {
      throw new VError(`Invalid number in range: ${parts[0]}`);
    }

    // If it's a single number, both start and end are the same
    if (parts.length === 1) {
      return [start, start];
    }

    const end = parseInt(parts[1]);
    if (isNaN(end)) {
      throw new VError(`Invalid number in range: ${parts[1]}`);
    }

    if (end < start) {
      throw new VError(`Invalid range ${range}: end cannot be less than start`);
    }

    return [start, end];
  }

  next(bucketId: number): number {
    let low = 0;
    let high = this.sortedBuckets.length - 1;
    while (low <= high) {
      const mid = Math.floor((low + high) / 2);
      if (this.sortedBuckets[mid] <= bucketId) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }
    if (low < this.sortedBuckets.length) {
      return this.sortedBuckets[low];
    }
    // Wrap around to first bucket if no larger bucket found
    return this.sortedBuckets[0];
  }
}
