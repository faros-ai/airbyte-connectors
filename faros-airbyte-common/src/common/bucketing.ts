import {createHmac} from 'crypto';
import VError from 'verror';

export interface BucketExecutionState {
  __bucket_execution_state?: {
    last_executed_bucket_id: number;
  };
  [key: string]: any;
}

export interface RoundRobinConfig {
  round_robin_bucket_execution?: boolean;
  bucket_id?: number;
  bucket_total?: number;
  [key: string]: any;
}

export function bucket(key: string, data: string, bucketTotal: number): number {
  const md5 = createHmac('md5', key);
  md5.update(data);
  const hex = md5.digest('hex').substring(0, 8);
  return (parseInt(hex, 16) % bucketTotal) + 1; // 1-index for readability
}

export function validateBucketingConfig(
  bucketId: number = 1,
  bucketTotal: number = 1
): void {
  if (bucketTotal < 1) {
    throw new VError('bucket_total must be a positive integer');
  }
  if (bucketId < 1 || bucketId > bucketTotal) {
    throw new VError(`bucket_id must be between 1 and ${bucketTotal}`);
  }
}

export function nextBucketId(
  config: RoundRobinConfig,
  state?: BucketExecutionState
): number {
  const bucketTotal = config.bucket_total ?? 1;
  const lastExecutedBucketId =
    state?.__bucket_execution_state?.last_executed_bucket_id ?? bucketTotal;

  return (lastExecutedBucketId % bucketTotal) + 1;
}

export function applyRoundRobinBucketing(
  config: RoundRobinConfig,
  state?: BucketExecutionState,
  logger?: (message: string) => void
): {config: RoundRobinConfig; state: BucketExecutionState} {
  if (!config.round_robin_bucket_execution) {
    return {config, state};
  }

  const next = nextBucketId(config, state);
  logger?.(`Using round robin bucket execution. Bucket id: ${next}`);

  return {
    config: {
      ...config,
      bucket_id: next,
    },
    state: {
      ...state,
      __bucket_execution_state: {
        last_executed_bucket_id: next,
      },
    },
  };
}
