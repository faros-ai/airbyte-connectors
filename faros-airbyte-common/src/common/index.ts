// TODO: Try https://www.npmjs.com/package/diff
export interface FileDiff {
  deletions: number;
  additions: number;
  from?: string;
  to?: string;
  deleted?: boolean;
  new?: boolean;
}
