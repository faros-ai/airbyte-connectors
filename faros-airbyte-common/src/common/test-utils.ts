import fs from 'fs';
import { readTestResourceFile, readTestResourceAsJSON } from 'faros-airbyte-cdk';

/**
 * Read a resource file from the specified path.
 * This function is compatible with the previous duplicated implementations
 * and maintains backward compatibility with resources in the 'resources/' directory.
 */
export function readResourceFile(fileName: string): any {
  return JSON.parse(fs.readFileSync(`resources/${fileName}`, 'utf8'));
}

/**
 * Parse a test resource from the 'test/resources/' directory
 * This is simply re-exporting the CDK function for convenience.
 */
export const readTestResource = readTestResourceFile;

/**
 * Parse a test resource into JSON from the 'test/resources/' directory
 * This is simply re-exporting the CDK function for convenience.
 */
export const readTestResourceJSON = readTestResourceAsJSON;
