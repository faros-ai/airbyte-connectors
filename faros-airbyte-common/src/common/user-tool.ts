import fs from 'fs';
import path from 'path';

/**
 * GraphQL query to fetch vcs_UserTool records from Faros.
 * Can be used to query for active/inactive user tools by source, organization, and tool category.
 * The toolDetail parameter is optional and can be omitted if not needed.
 */
export const VCS_USER_TOOL_QUERY = fs.readFileSync(
  path.join(
    __dirname,
    '..',
    '..',
    'resources',
    'common',
    'queries',
    'vcs-user-tool.gql'
  ),
  'utf8'
);
