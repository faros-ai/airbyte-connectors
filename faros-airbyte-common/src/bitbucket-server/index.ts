// TODO: Deduplicate shared types for other source/converter pairs

import {FileDiff} from '../common';

type HRefs = {self?: {href: string}[]};
export function selfHRef(links: HRefs): string | undefined {
  return links.self?.find((l) => l.href)?.href;
}

export interface Commit {
  readonly id: string;
  readonly author: User;
  readonly authorTimestamp: number;
  readonly committer: User;
  readonly committerTimestamp: number;
  readonly message: string;
  readonly computedProperties: {
    readonly repository: {readonly fullName: string};
  };
}

export interface PullRequestActivity {
  readonly id: number;
  readonly createdDate: number;
  readonly user: User;
  readonly action: string;
  readonly computedProperties: {
    readonly pullRequest: {
      readonly id: number;
      readonly repository: {readonly fullName: string};
      readonly updatedDate: number;
    };
  };
}

interface PullRequestComment extends PullRequestActivity {
  readonly comment: {
    readonly text: string;
    readonly createdDate: number;
    readonly updatedDate: number;
  };
}

export function isPullRequestComment(
  activity: PullRequestActivity
): activity is PullRequestComment {
  return activity.action === 'COMMENTED';
}

interface PullRequestMerge extends PullRequestActivity {
  readonly commit: Commit;
}

export function isPullRequestMerge(
  activity: PullRequestActivity
): activity is PullRequestMerge {
  return activity.action === 'MERGED';
}

export function isPullRequestReview(activity: PullRequestActivity): boolean {
  return (
    activity.action === 'APPROVED' ||
    activity.action === 'DECLINED' ||
    activity.action === 'REVIEWED'
  );
}

export interface PullRequestDiff {
  readonly files: ReadonlyArray<FileDiff>;
  readonly computedProperties: {
    readonly pullRequest: {
      readonly id: number;
      readonly repository: {readonly fullName: string};
      readonly updatedDate: number;
    };
  };
}

export interface PullRequest {
  readonly author: {readonly user: User};
  readonly id: number;
  readonly title: string;
  readonly description: string;
  readonly state: string;
  readonly createdDate: number;
  readonly updatedDate: number;
  readonly toRef: {
    readonly repository: Repository;
  };
  readonly properties: {readonly commentCount: number};
  readonly links: HRefs;
  readonly computedProperties: {
    readonly repository: {readonly fullName: string};
  };
}

export interface Repository {
  readonly slug: string;
  readonly name: string;
  readonly description: string;
  readonly project: Project;
  readonly public: boolean;
  readonly links: HRefs;
  readonly computedProperties: {
    readonly fullName: string;
    readonly mainBranch: string;
  };
}

export interface Tag {
  readonly id: string;
  readonly displayId: string;
  readonly type: string;
  readonly latestCommit: string;
  readonly latestChangeset: string;
  readonly hash: string;
  readonly computedProperties: {
    readonly repository: {readonly fullName: string};
  };
}

export interface Project {
  readonly key: string;
  readonly name: string;
  readonly links: HRefs;
}

export interface User {
  readonly displayName: string;
  readonly emailAddress: string;
  readonly name: string;
  readonly slug: string;
  readonly links: HRefs;
}

export interface ProjectUser {
  readonly user: User;
  readonly project: {readonly key: string};
}
