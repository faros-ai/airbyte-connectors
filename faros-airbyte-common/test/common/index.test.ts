import {calculateDateRange} from '../../src/common';

describe('calculateDateRange', () => {
  const logger = jest.fn();

  beforeEach(() => {
    logger.mockClear();
  });

  it('should calculate date range with start and end date', () => {
    const result = calculateDateRange({
      start_date: '2022-01-01',
      end_date: '2022-01-31',
      logger,
    });

    expect(result).toMatchSnapshot();
    expect(logger).toHaveBeenCalledWith(
      `Will process data from ${result.startDate} to ${result.endDate}`
    );
  });

  it('should calculate date range with end date and cutoff days', () => {
    const result = calculateDateRange({
      end_date: '2022-01-31',
      cutoff_days: 10,
      logger,
    });

    expect(result).toMatchSnapshot();
    expect(logger).toHaveBeenCalledWith(
      'Cutoff days provided, calculating start date from end date'
    );
  });

  it('should calculate date range with current date as end date if not provided', () => {
    const result = calculateDateRange({
      start_date: '2022-01-01',
      logger,
    });

    expect(result.startDate).toMatchSnapshot();
    expect(result.endDate).toBeDefined();
    expect(logger).toHaveBeenCalledWith(
      'End date not provided, using current date'
    );
  });

  it('should throw error if start date is after end date', () => {
    expect(() =>
      calculateDateRange({
        start_date: '2022-02-01',
        end_date: '2022-01-31',
        logger,
      })
    ).toThrow(/Start date: .* is after end date: .*/);
  });

  it('should throw error if neither start date nor cutoff days are provided', () => {
    expect(() =>
      calculateDateRange({
        end_date: '2022-01-31',
        logger,
      })
    ).toThrow('Either start_date or cutoff_days must be provided');
  });

  it('should throw error if both start date and cutoff days are provided', () => {
    calculateDateRange({
      start_date: '2022-01-01',
      cutoff_days: 10,
      logger,
    });
    expect(logger).toHaveBeenCalledWith(
      'Both start date and cutoff days provided, discarding cutoff days'
    );
  });
});
