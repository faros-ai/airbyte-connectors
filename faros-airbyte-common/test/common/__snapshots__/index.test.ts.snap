// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateDateRange should calculate date range with current date as end date if not provided 1`] = `2022-01-01T00:00:00.000Z`;

exports[`calculateDateRange should calculate date range with end date and cutoff days 1`] = `
Object {
  "endDate": 2022-01-31T00:00:00.000Z,
  "startDate": 2022-01-21T00:00:00.000Z,
}
`;

exports[`calculateDateRange should calculate date range with start and end date 1`] = `
Object {
  "endDate": 2022-01-31T00:00:00.000Z,
  "startDate": 2022-01-01T00:00:00.000Z,
}
`;
